File: Models/RepeatMode.swift
import Foundation

/// Режимы повтора.
/// - off: без повтора
/// - one: повтор текущего трека
/// - all: повтор всего плейлиста
enum RepeatMode: String, CaseIterable, Codable, Identifiable {
    case off = "off"
    case one = "one"
    case all = "all"

    var id: String { rawValue }

    /// Стабильный локализационный ключ, не зависящий от сохраненного rawValue.
    var localizedKey: String {
        switch self {
        case .off:
            return "repeat.off"
        case .one:
            return "repeat.one"
        case .all:
            return "repeat.all"
        }
    }

    static func fromStoredValue(_ value: String) -> RepeatMode? {
        if let mode = RepeatMode(rawValue: value) {
            return mode
        }

        // Backward compatibility for pre-localization persisted raw values.
        switch value {
        case "Без повтора":
            return .off
        case "Повтор трека":
            return .one
        case "Повтор плейлиста":
            return .all
        default:
            return nil
        }
    }
}

File: windows/src/DungeonSoundboard.Core/Models/RepeatMode.cs
using System.Text.Json.Serialization;

namespace DungeonSoundboard.Core.Models;

[JsonConverter(typeof(JsonStringEnumConverter<RepeatMode>))]
public enum RepeatMode
{
    [JsonStringEnumMemberName("off")]
    Off,

    [JsonStringEnumMemberName("one")]
    One,

    [JsonStringEnumMemberName("all")]
    All
}

public static class RepeatModeExtensions
{
    public static string LocalizedKey(this RepeatMode mode)
    {
        return mode switch
        {
            RepeatMode.Off => "repeat.off",
            RepeatMode.One => "repeat.one",
            RepeatMode.All => "repeat.all",
            _ => "repeat.off"
        };
    }

    public static RepeatMode? FromStoredValue(string value)
    {
        return value switch
        {
            "off" => RepeatMode.Off,
            "one" => RepeatMode.One,
            "all" => RepeatMode.All,
            "Без повтора" => RepeatMode.Off,
            "Повтор трека" => RepeatMode.One,
            "Повтор плейлиста" => RepeatMode.All,
            _ => null
        };
    }
}
